docker/docker.ts

```diff
@@ -175,36 +180,76 @@ async function buildAndPushImageWorkerAsync(
+        // running the pull if the repository had not yet been created.
         const cacheFromParam = typeof pathOrBuild.cacheFrom === "boolean" ? {} : pathOrBuild.cacheFrom;
-        cacheFrom = pullCacheAsync(imageName, cacheFromParam, login, repositoryUrl, logResource);
-    } else {
-        cacheFrom = Promise.resolve(undefined);
+        cacheFrom = await pullCacheAsync(imageName, cacheFromParam, login, repositoryUrl, logResource);
     }
 
     // First build the image.
-    const buildResult = await buildImageAsync(imageName, pathOrBuild, logResource, cacheFrom);
+    const { imageId, stages } = await buildImageAsync(imageName, pathOrBuild, logResource, cacheFrom);
+
+    if (!repositoryUrl) {
+        throw new ResourceError("Expected repository URL to be defined during push", logResource);
+    }
 
-    // Use those then push the image.  Then just return the digest as the final result for our caller to use.
+    // Generate a name that uniquely will identify this built image.  This is similar in purpose to
+    // the name@digest form that can be normally be retrieved from a docker repository.  However,
+    // this tag doesn't require actually pushing the image, nor does it require communicating with
+    // some external system, making it suitable for unique identification, even during preview.
+    // This also means that if docker produces a new imageId, we'll get a new name here, ensuring that
+    // resources (like docker.Image and cloud.Service) will be appropriately replaced.
+    const uniqueTargetName = createTargetName(repositoryUrl, tag, imageId);
+
+    // Use those then push the image.  Then just return the unique target name. as the final result
+    // for our caller to use.
     if (!pulumi.runtime.isDryRun()) {
-        // Only push the image during an update, do not push during a preview, even if digest and url are available
-        // from a previous update.
+        // Only push the image during an update, do not push during a preview.
         await login();
 
         // Push the final image first, then push the stage images to use for caching.
-        await pushImageAsync(imageName, repositoryUrl, logResource);
+        await tagAndPushImageAsync(imageName, repositoryUrl, tag, imageId, logResource);
 
-        for (const stage of buildResult.stages) {
-            await pushImageAsync(
-                localStageImageName(imageName, stage), repositoryUrl, logResource, stage);
+        for (const stage of stages) {
+            await tagAndPushImageAsync(
+                localStageImageName(imageName, stage), repositoryUrl, tag, /*imageId:*/ undefined, logResource);
         }
     }
 
-    return buildResult.id;
+    return uniqueTargetName;
+}
+
+function localStageImageName(imageName: string, stage: string) {
+    return `${imageName}-${stage}`;
+}
+
+function getImageNameAndTag(baseImageName: string): { imageName: string, tag: string | undefined } {
+    const lastColon = baseImageName.lastIndexOf(":");
```

Comment:
It's not 100% clear from https://docs.docker.com/engine/reference/commandline/tag/ that this is valid parsing - that is - that the tag is not allowed to contain a `:`.  It's not actually clear from there what the unambiguous grammar of these names is.  This is probably okay - but would love to have a reference to a grammar.

Reply:
The docs are here: https://docs.docker.com/registry/spec/api/#content-digests
I'll link in the PR.  They say:

> To disambiguate from other concepts, we call this identifier a digest. A digest is a serialized hash result, consisting of a algorithm and hex portion. The algorithm identifies the methodology used to calculate the digest. The hex portion is the hex-encoded result of the hash.

So we're good here.

Reply:
oh sorry, this was in referece to the tag.  let me find the docs for that

Reply:
Docs say this:

> A tag name must be valid ASCII and may contain lowercase and uppercase letters, digits, underscores, periods and dashes. A tag name may not start with a period or a dash and may contain a maximum of 128 characters.

https://docs.docker.com/engine/reference/commandline/tag/